### PR TITLE
week-6

### PR DESCRIPTION
**Pull request by oleg3555**
Add settings, switch of theme, used context API, add loading spinner, add Alerts.
Login: oleg.yanysik@gmail.com
Password: qwer1234

[Link to deploy](https://roaring-sawine-6b7d51.netlify.app/login)

### CODE REVIEW

File: src/App/App.jsx [hunk 1]
```diff
@@ -15,19 +16,23 @@ class App extends PureComponent {
...
-            </Switch>
-          </ErrorBoundary>
-        </main>
-        <footer>
-          <span className={styles.copyright}>{COPYRIGHT}</span>
-        </footer>
-      </div>
+      <ThemeContext.Consumer>
+        {({ theme }) => (
+          <div className={styles.App} style={{ backgroundColor: theme.backgroundColor }}>
```
Comment (maximsan) [inline src/App/App.jsx:21]: this approach is not good, use class for theme or find another approach but not this one 😉

**Review by maximsan — CHANGES_REQUESTED** [review]
Hello! I left some comments. BTW where is redux integration or I missed something? ))

File: package.json [hunk 2]
```diff
@@ -118,6 +118,8 @@
     "cz-conventional-changelog": "3.3.0",
     "eslint-config-airbnb": "19.0.4",
     "eslint-config-prettier": "8.3.0",
+    "eslint-config-react": "^1.1.7",
+    "eslint-config-react-app": "^7.0.1",
```
[the remark this answers is not in the document]
Comment (oleg3555) [reply package.json:125]: It was eslint errors until I add this

File: src/App/App.jsx (hunk 1 above)
Comment (oleg3555) [reply src/App/App.jsx:21]: Fixed

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (maximsan) [reply package.json:125]: it's because you installed dependencies wrongly, but let's skip it for now

File: src/App/GeneratedRoutes.jsx [hunk 3]
```diff
@@ -17,11 +20,17 @@ export function GeneratedRoutes() {
...
+              {user.role === USER_ROLES.admin || user.role === USER_ROLES.mentor ? (
+                <>
+                  <Route path='/users' exact render={() => <Members store={store} />} />
+                  <Route path='/tasks' exact render={() => <Tasks store={store} />} />
+                  <Route path='/progress/:id' component={Progress} />
+                </>
+              ) : (
+                <Route path='/track/:userId/task/:taskId' component={Tracks} />
+              )}
+              <Route path='/tasks/:id' render={({ match }) => <UserTasks match={match} />} />
```
Comment (maximsan) [inline src/App/GeneratedRoutes.jsx:32]: You have access to match by default inside props. Why do you do it like this ? There is a mistake somewhere in your integration

File: src/components/Buttons/Button/Button.jsx [hunk 4]
```diff
@@ -1,10 +1,30 @@
...
+    error: 'var(--red)',
+  },
+  dark: {
+    primary: 'var(--darkBlue)',
+    warning: 'var(--darkOrange)',
+    success: 'var(--darkGreen)',
+    secondary: 'var(--secondary)',
+    error: 'var(--darkRed)',
+  },
+};
```
Comment (maximsan) [inline src/components/Buttons/Button/Button.jsx:21]: You overcomplicate it here. Think how to simplify it via css classes.

File: src/pages/modals/form/ModalFields/Input/Input.jsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+import styles from './Input.module.css';
+import fieldStyles from '../ModalFields.module.css';
+import noop from '../../../../../shared/noop';
+import { INPUT_TYPES } from '../../../../../constants/libraries';
+import { Error } from '../../../../../components/Error/Error';
+import { ThemeContext } from '../../../../../providers/ThemeProvider';
+
+export function Input({ onChange, value, placeholder, readOnly, fieldName, error, type, title }) {
+  const onChangeHandler = ({ target }) => {
+    onChange(target.name, target.value);
```
Comment (maximsan) [inline src/pages/modals/form/ModalFields/Input/Input.jsx:11]: scpread target

File: src/pages/tasks/Tasks.jsx [hunk 6]
```diff
@@ -1,153 +1,160 @@
...
     );
   }
 }
 
+function mapStateToProps(state) {
+  return {
+    users: state.users,
+    tasks: state.tasks,
+    isFetching: state.fetch.isFetching,
+    error: state.fetch.error,
```
Comment (maximsan) [inline src/pages/tasks/Tasks.jsx:112]: `fetch` field is redundant, make it simple

File: src/pages/tasks/Tasks.jsx [hunk 7]
```diff
@@ -1,153 +1,160 @@
...
+  };
+}
+
+function mapDispatchToProps(dispatch) {
+  return {
+    getUsers: () => dispatch(getUsersThunk()),
+    getTasks: () => dispatch(getTasksThunk()),
+    updateTask: (id, updatedTask) => dispatch(updateTaskThunk(id, updatedTask)),
+    removeTask: (id) => dispatch(removeTaskThunk(id)),
+    addTask: (task) => dispatch(addTaskThunk(task)),
```
Comment (maximsan) [inline src/pages/tasks/Tasks.jsx:122]: use `bindActionCreators` to concise your code

File: src/redux/fetchReducer/fetchReducer.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    case DISABLE_ERROR: {
+      return { ...state, error: '' };
+    }
+    default: {
+      return state;
+    }
+  }
+};
+
+export const enableLoaderAC = () => ({ type: ENABLE_LOADER });
```
Comment (maximsan) [inline src/redux/fetchReducer/fetchReducer.js:29]: `AC` part in name looks weird, I'd rather recommend to not add it at all

File: src/redux/tasksReducer/tasksReducer.js [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+export const tasksReducer = (state = initialState, action) => {
+  switch (action.type) {
+    case GET_TASKS: {
+      return action.payload.tasks;
+    }
+    case UPDATE_TASK: {
+      return state.map((task) => (task.id === action.payload.id ? { ...task, ...action.payload.task } : task));
+    }
+    case ADD_TASK: {
+      return [...state, { id: action.payload.id, ...action.payload.task }];
```
Comment (maximsan) [inline src/redux/tasksReducer/tasksReducer.js:15]: just pass a task object

File: src/redux/tasksThunk/tasksThunk.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
+import { disableErrorAC, disableLoaderAC, enableLoaderAC, setErrorAC } from '../fetchReducer/fetchReducer';
+import { addTask, deleteTask, deleteTrack, getAllTasks, updateTask } from '../../scripts/api-service';
+import { addTaskAC, getTasksAC, removeTaskAC, updateTaskAC } from '../tasksReducer/tasksReducer';
+
+export const getTasksThunk = () => {
+  return async (dispatch) => {
+    dispatch(enableLoaderAC());
+    const tasks = await getAllTasks();
```
Comment (maximsan) [inline src/redux/tasksThunk/tasksThunk.js:8]: do you handle possible errors ?

File: src/redux/tasksThunk/tasksThunk.js [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    }
+    dispatch(disableLoaderAC());
+  };
+};
+
+export const addTaskThunk = (task) => {
+  return (dispatch) => {
+    dispatch(enableLoaderAC());
+    addTask(task)
+      .then((doc) => dispatch(addTaskAC(doc.id, task)))
```
Comment (maximsan) [inline src/redux/tasksThunk/tasksThunk.js:49]: Let's stick to one way. Please use `Promise chaining` or `async wait` everywhere.

File: src/redux/usersReducer/usersReducer.js [hunk 12]
```diff
@@ -0,0 +1,29 @@
+import { ADD_USER, GET_USERS, REMOVE_USER, UPDATE_USER } from './type-constants';
+
+const initialState = [];
+
+// eslint-disable-next-line default-param-last
```
Comment (maximsan) [inline src/redux/usersReducer/usersReducer.js:5]: Do you need it ? It's better to change you code to avoid disabling of extra eslint rules

**Review by maximsan — CHANGES_REQUESTED** [review]
Hello! I left some comments. Please fix them. The most urgent is the store using, please pay attention 😉

File: src/App/GeneratedRoutes.jsx (hunk 3 above)
Comment (oleg3555) [reply src/App/GeneratedRoutes.jsx:32]: Fixed

File: src/App/GeneratedRoutes.jsx [hunk 13]
```diff
@@ -17,11 +20,17 @@ export function GeneratedRoutes() {
...
             <>
-              <Route path='/users' exact component={Members} />
-              <Route path='/tasks' exact component={Tasks} />
-              <Route path='/progress/:id' component={Progress} />
-              <Route path='/tasks/:id' component={UserTasks} />
-              <Route path='/track/:userId/task/:taskId' component={Tracks} />
+              {user.role === USER_ROLES.admin || user.role === USER_ROLES.mentor ? (
+                <>
+                  <Route path='/users' exact render={() => <Members store={store} />} />
+                  <Route path='/tasks' exact render={() => <Tasks store={store} />} />
```
[the remark this answers is not in the document]
Comment (oleg3555) [reply src/App/GeneratedRoutes.jsx:26]: Fixed

File: src/pages/modals/form/ModalFields/Input/Input.jsx (hunk 5 above)
Comment (oleg3555) [reply src/pages/modals/form/ModalFields/Input/Input.jsx:11]: Fixed

File: src/redux/tasksThunk/tasksThunk.js (hunk 10 above)
Comment (oleg3555) [reply src/redux/tasksThunk/tasksThunk.js:8]: It is try catch block inside `getAllTasks` method

File: src/redux/usersReducer/usersReducer.js (hunk 12 above)
Comment (oleg3555) [reply src/redux/usersReducer/usersReducer.js:5]:
```suggestion
export const tasksReducer = (state = initialState, action = {}) => {
```
Fixed

File: src/redux/fetchReducer/fetchReducer.js (hunk 8 above)
Comment (oleg3555) [reply src/redux/fetchReducer/fetchReducer.js:29]: Fixed

File: src/components/Buttons/Button/Button.jsx (hunk 4 above)
Comment (oleg3555) [reply src/components/Buttons/Button/Button.jsx:21]: Fixed

File: src/redux/tasksThunk/tasksThunk.js (hunk 11 above)
Comment (oleg3555) [reply src/redux/tasksThunk/tasksThunk.js:49]: Rewrote to async await

File: src/pages/tasks/Tasks.jsx (hunk 6 above)
Comment (oleg3555) [reply src/pages/tasks/Tasks.jsx:112]: Fixed

**Review by maximsan — APPROVED** [review]
Hi! Well done. Approved!

**Review by LeoKuD — APPROVED** [review]
Good job. Have a good week.

**Pull request merged**